tensorboard/data/server/tf_record.rs

```diff
@@ -0,0 +1,327 @@
+        let want = self.data_crc;
+        if got == want {
+            Ok(())
+        } else {
+            Err(ChecksumError { got, want })
+        }
+    }
+}
+
+/// Error returned by [`TfRecordState::read_record`].
+#[derive(Debug)]
+pub enum ReadRecordError {
+    /// Length field failed checksum. The file is corrupt, and reading must abort.
+    BadLengthCrc(ChecksumError),
+    /// No fatal errors so far, but the record is not complete. Call `read_record` again with the
+    /// same state buffer once new data may be available.
+    ///
+    /// This includes the "trivial truncation" case where there are no bytes in a new record, so
+    /// repeatedly reading records from a file of zero or more well-formed records will always
+    /// finish with a `Truncated` error.
+    Truncated,
+    /// Record is too large to be represented in memory on this system.
+    ///
+    /// In principle, it would be possible to recover from this error, but in practice this should
+    /// rarely occur since serialized protocol buffers do not exceed 2 GiB in size. Thus, no
+    /// recovery codepath has been implemented, so reading must abort.
+    TooLarge(u64),
+    /// Underlying I/O error. May be retryable if the underlying error is.
+    Io(io::Error),
+}
+
+impl From<io::Error> for ReadRecordError {
+    fn from(e: io::Error) -> Self {
+        ReadRecordError::Io(e)
+    }
+}
+
+impl TfRecordState {
+    /// Attempts to read a TFRecord, pausing gracefully in the face of truncations. If the record
+    /// is truncated, the result is a `Truncated` error, and the state buffer will be updated to
+    /// contain the prefix of the raw record that was read. The same state buffer should be passed
+    /// to a subsequent call to `read_record` that it may continue where it left off. If the record
+    /// is read successfully, this `TfRecordState` is left at its default value (equivalent to
+    /// `TfRecordState::new`, but without re-allocating) and may be reused by the caller to read a
+    /// fresh record.
+    ///
+    /// The record's length field is always validated against its checksum, but the full data is
+    /// only validated if you call `checksum()` on the resulting record.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use rustboard_core::tf_record::{ReadRecordError, TfRecordState};
+    /// use std::io::Cursor;
+    ///
+    /// let mut buf: Vec<u8> = Vec::new();
+    /// buf.extend(b"\x18\x00\x00\x00\x00\x00\x00\x00"); // length: 24 bytes
+    /// buf.extend(b"\xa3\x7f\x4b\x22"); // length checksum (0x224b7fa3)
+    /// let contents = b"\x09\x00\x00\x80\x38\x99\xd6\xd7\x41\x1a\x0dbrain.Event:2";
+    /// buf.extend(&contents[..5]); // file truncated mid-write
+    ///
+    /// let mut state = TfRecordState::new();
+    ///
+    /// // First attempt: read what we can, then encounter truncation.
+    /// assert!(matches!(
+    ///     state.read_record(&mut Cursor::new(buf)),
```

Comment:
> I must confess that I am not used to thinking data first; I find Rust
> particularly interesting since you think data struct first then
> traits, then method/impls later (exact ordering is not important but
> emphasis on thinking about type flowing before class in OOP).

Definitely. This is a good description.

> Now, I find it weird for RecordState to be passed with a buffer that
> changes its internal state. Since I am definitely new to this, I would
> value your opinion/code a lot more, but I just want to make sure why
> it is wrong to have, for example, something like TfRecordReader
> instead. Is current code more natural when thinking the data/state
> first?

This is also acceptable! A `TfRecordReader` could take an underlying
file reader as an argument and own it, and then just expose a
`read_record(&mut self)` method that calls the underlying reader.

For instance, `BufReader` in the standard library does this:
<https://doc.rust-lang.org/std/io/struct.BufReader.html#method.new>

If you let a `TfRecordReader` own and wrap an underlying reader, then
you have an additional guarantee: nothing else can access the underlying
reader anymore. For instance, in Java, you could write this:

```java
FileReader fr = new FileReader("/etc/hosts");
BufferedReader br = new BufferedReader(br);
br.readLine();
fr.read();  // poke into the underlying reader!
br.readLine();  // undefined behavior, probably! def. not what you want
```

…but in Rust [that’s a borrow-checker error][1]:

```rust
use std::fs::File;
use std::io::{BufRead, BufReader, Read};

fn main() -> std::io::Result<()> {
    let mut fr: File = File::open("/etc/hosts")?;
    let mut br: BufReader<File> = BufReader::new(fr);
    let mut line = String::new();
    br.read_line(&mut line)?;
    dbg!(line);
    fr.read(&mut [])?;
    Ok(())
}
```

```
error[E0382]: borrow of moved value: `fr`
  --> src/main.rs:10:5
   |
5  |     let mut fr: File = File::open("/etc/hosts")?;
   |         ------ move occurs because `fr` has type `std::fs::File`, which does not implement the `Copy` trait
6  |     let mut br: BufReader<File> = BufReader::new(fr);
   |                                                  -- value moved here
...
10 |     fr.read(&mut [])?;
   |     ^^ value borrowed here after move
```

For a `BufReader`, this is definitely a good thing, since it keeps an
internal buffer of extra data read from the file, and reading directly
from the underlying stream would break its invariants. What about for a
`TfRecordReader`? It feels like the same thing is true: if you just read
ahead in the file, then our bookkeeping about where in the record layout
we are will be incorrect. Preventing that is nice.

So, sure! Let’s let a `TfRecordReader` own an underlying reader.

There’s one place in our code where we _do_ want to poke the underlying
reader, though—in the documentation example for `read_record`. See: we
call `read_record(&mut Cursor::new(buf))`, then we change `buf`, then we
call it again with `&mut Cursor::new(buf)`, which is a different reader.
So one approach would be to expose

```rust
impl<R> TfRecordReader<R> {
    /// Borrows the underlying reader.
    pub fn as_inner(&mut self) -> &mut R {
        &mut self.reader
    }
}
```

which still lets our reader own the underlying reader but enables
callers to borrow it. Of course, if they borrow it, they can muck with
its state again, so we lose some of the guarantee that we’ve gained. So
instead what I’ve done is work around it a bit in the example, by
defining a reader type that actually does grow, just like a file does.
I’ve used [leading `#` comments][dt] to hide this part from the docs, so
that it still powers the doctest but doesn’t distract readers. You can
check out the doc server, if you like.

What do you think?

[1]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bd57b5c90cd06bcc697cf355698b20cd
[dt]: https://doc.rust-lang.org/rustdoc/documentation-tests.html#hiding-portions-of-the-example

Reply:
> I find Rust particularly interesting since you think data struct first
> then traits, then method/impls later

You may also be interested in [this koan, “Behavioural modeling”][bm]:
<https://users.rust-lang.org/t/rust-koans/2408/2>

I didn’t fully understand it the first time I read it. Later, while
doing something entirely unrelated (in JavaScript!), I was enlightened.

[bm]: https://users.rust-lang.org/t/rust-koans/2408/2